onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -572,41 +580,125 @@ class PlannerImpl {
+    const KernelCreateInfo& kernel_create_info = GetKernelCreateInfo(kernel_create_info_map, node.Index());
 
     if (utils::IsInputOnCpu(node, &kernel_create_info, input_index))
       // weights are not output from any node, so it's OK to put its location on CPU provider
-      return execution_providers_.GetDefaultCpuMemoryInfo();
+      return execution_providers.GetDefaultCpuMemoryInfo();
     return p_provider->GetAllocator(0, OrtMemTypeDefault)->Info();
   }
 
-  Status GeneratePlanForWeights() {
-    auto& weights = graph_viewer_.GetAllInitializedTensors();
-    std::vector<std::vector<OrtMemoryInfo>> locations(plan_.allocation_plan.size());
-    for (const auto& node : graph_viewer_.Nodes()) {
-      auto status = onnxruntime::Node::ForEachWithIndex(
-          node.InputDefs(), [this, &locations, &node, &weights](const onnxruntime::NodeArg& def, size_t index) {
-            auto sub_status = Status::OK();
-            ORT_TRY {
-              auto& def_name = def.Name();
-              if (!weights.count(def_name)) return Status::OK();
-              auto wt_index = Index(def_name);
-              locations[wt_index].emplace_back(GetLocationForNodeInput(index, node));
+  Status GeneratePlanForWeightsHelper(const GraphViewer& graph_viewer,
+                                      const InitializedTensorSet& weights,
+                                      const KernelCreateInfoMap& kernel_create_info_map,
+                                      const ExecutionProviders& execution_providers,
+                                      const std::string& subgraph_kernel_create_info_map_key_base,
+                                      size_t graph_depth,
+                                      /*out*/ std::vector<std::vector<OrtMemoryInfo>>& locations) {
+    for (const auto& node : graph_viewer.Nodes()) {
+      const auto& input_node_args = node.InputDefs();
+      size_t num_node_inputs = input_node_args.size();
+
+      for (size_t node_input_index = 0; node_input_index < num_node_inputs; ++node_input_index) {
+        auto input_node_arg = input_node_args[node_input_index];
+
+        // Skip processing missing optional inputs
+        if (!input_node_arg->Exists()) {
+          continue;
+        }
+
+        ORT_TRY {
+          auto& def_name = input_node_arg->Name();
+
+          // This node input doesn't correspond to any of the weights
+          if (!weights.count(def_name)) {
+            continue;
+          }
+
+          // Skip processing shadow values in subgraphs.
+          // While processing subgraphs, if we don't see this entry in the implicit
+          // inputs of the node containing the subgraph, skip processing it as it is
+          // a shadow value.
+          if (graph_depth > 0) {
+            // We are processing a subgraph if we enter this
+            const auto* parent_node = graph_viewer.ParentNode();
+            ORT_ENFORCE(parent_node);
+
```

Comment:
Do we need this? Costs binary size and really this should be a) impossible and b) validated elsewhere.